tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,269 @@
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  T return_val = *dest;
+  *dest = value;
+  //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  //unlock_address_openacc((void*)dest, scope);
+  return return_val;
+}
+
+#ifdef KOKKOS_COMPILER_NVHPC
+  //FIXME_OPENACC_NVHPC: Old NVHPC fails in compiling atomic 
+  //capture constructs in some cases, and thus CUDA intrinsic 
+  //function (atomicExch) is used instead.
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_arithmetic<T>::value && (sizeof(T) != 4) && (sizeof(T) != 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope scope) {
+  if (acc_on_device(acc_device_not_host)) {
+    printf("DESUL error in device_atomic_exchange(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  T return_val = *dest;
+  *dest = value;
+  //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  //unlock_address_openacc((void*)dest, scope);
+  return return_val;
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 4), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(unsigned int) == 4,
+                "this function assumes an unsigned int is 32-bit");
+  unsigned int return_val = atomicExch(reinterpret_cast<unsigned int*>(dest),
+                                       reinterpret_cast<unsigned int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(unsigned long long int) == 8,
+                "this function assumes an unsigned long long int is 64-bit");
+  unsigned long long int return_val = atomicExch(reinterpret_cast<unsigned long long int*>(dest),
+                                       reinterpret_cast<unsigned long long int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 4), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(float) == 4,
+                "this function assumes an float is 32-bit");
+  float return_val = atomicExch(reinterpret_cast<float*>(dest),
+                                       reinterpret_cast<float&>(value));
+  return reinterpret_cast<T&>(return_val);
```

Comment:
Compiler bug?

Reply:
It seems to be a compile bug that happens only when `reinterpret_cast` is used together with `atomicExch()`.
By the way, the implementation of `device_atomic_exchange()` is changed not to use `atomicExch()`; thus this issue does not occur in `device_atomic_exchange()`. 
However, a similar compiler bug still occurs in `device_atomic_compare_exchange()` because this API uses both `reinterpret_cast` and `atomicCAS()` together.